torch/distributed/_sharded_tensor/api.py

```diff
@@ -335,6 +352,54 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._all_gather(None)
 
+    @classmethod
+    def _gather(
+        cls,
+        sharded_tensor: "ShardedTensor",
+        dst_rank: int = 0,
+        dst_tensor: Optional[torch.Tensor] = None,
+    ):
+        my_rank = dist.get_rank(sharded_tensor._process_group)
+        _validate_output_tensor_for_rank(my_rank, dst_rank, dst_tensor)
+
+        shard_tensors = sharded_tensor.local_shards
+
+        gathered_shards = [None for _ in sharded_tensor._process_group.world_size]
+
+        dist.gather_object(
+            obj=shard_tensors,
+            object_gather_list=gathered_shards,
+            dst=dst_rank,
+            group=sharded_tensor._process_group,
+        )
+
+        if my_rank == dst_rank:
+            for sharded_tensor in gathered_shards:
+                sizes = sharded_tensor.metadata().size
+                if sizes != dst_tensor.size():
+                    raise ValueError(
+                        f"dst_tensor need to have the same size as metadata.size {sizes}"
+                    )
+                dims = len(sizes)
+                pad = [0] * 2 * dims
+                for dim in range(dims):
```

Comment:
Are we padding each individual shard to be the size of the actual tensor and then just adding it? This is probably wasteful from a memory perspective. Can't we just get a view of `dst_tensor` based on offsets and lengths and just copy the shard into the appropriate position? You can probably use narrow (https://pytorch.org/docs/stable/generated/torch.narrow.html) multiple times to get this?

Reply:
I discussed this as well in the internal diff. let me know your thoughts